guardian/prosemirror-embeds
# Remove Fields (flat state) in favour of Props (nested state), which are now called ... Fields

## What does this change?

This PR gets rid of the last vestiges of the previous sort of state management (a single, empty parent node, with lots of fields), in favour of the current sort of state management (one to many child nodes, which have content or fields).

We now display the current values of the fields, and their validation output, to show that everything's working as intended:

<img width="808" alt="Screenshot 2021-05-17 at 17 38 05" src="https://user-images.githubusercontent.com/7767575/118525033-a936f400-b736-11eb-9754-cbb8cc0516a2.png">

In doing the above, I've tried to 
- keep things as typesafe as possible. As a result, the validation and field values passed to the consumer code are now typesafe, with tests
- tried to name things sensibly, and comment the more complicated type wrangling that's needed for the above.

As a result, this diff is larger than I intended, mainly because there's lots of renaming – v. happy to pair on reviewing.

There's one bug, which is longstanding but only visible now – initial values aren't reflected in the field values until the editor state changes. That'll have to wait for another PR.

## How to test

- The automated tests should pass. The new ones should make us feel confident we've got the types right
- The names for types should make at least _some_ sense. Lmk if things need renaming, or bear some explanation w/ comments.
